subsys/bluetooth/host/hci_core.c

```diff
@@ -3638,6 +3638,23 @@ static void hci_vs_init(void)
 static int hci_init(void)
 {
 	int err;
+
+	if (bt_dev.id_count > 0 && bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
+#if defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR)
+		if (bt_dev.drv->set_public_addr != NULL) {
+			err = bt_dev.drv->set_public_addr(&bt_dev.id_addr[BT_ID_DEFAULT].a);
+			if (err) {
+				LOG_ERR("Failed to set public address (%d)", err);
+				return err;
+			}
+		} else
+#endif /* defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR) */
+		{
+			LOG_ERR("Setting public address not supported by driver");
+			return -ENOTSUP;
+		}
+	}
```

Comment:
This can be refactored so that the ifdefs "fit into the parse tree" and look like true ifs. It think that's much more preferable.

```suggestion
	if (bt_dev.id_count > 0 && bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
#if defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR)
		if (bt_dev.drv->set_public_addr != NULL) {
			err = bt_dev.drv->set_public_addr(&bt_dev.id_addr[BT_ID_DEFAULT].a);
			if (err) {
				LOG_ERR("Failed to set public address (%d)", err);
			}
			return err;
		}
#endif /* defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR) */

		LOG_ERR("Setting public address not supported by driver");
		return -ENOTSUP;
	}
```

I also have a preference for using the IF_ENABLED macro instead of the ifdefs in these cases, as this gets a more readable indentation.

```suggestion
	if (bt_dev.id_count > 0 && bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
		IF_ENABLED(CONFIG_BT_HCI_SET_PUBLIC_ADDR, ({
			if (bt_dev.drv->set_public_addr != NULL) {
				err = bt_dev.drv->set_public_addr(&bt_dev.id_addr[BT_ID_DEFAULT].a);
				if (err) {
					LOG_ERR("Failed to set public address (%d)", err);
				}
				return err;
			}
		})

		LOG_ERR("Setting public address not supported by driver");
		return -ENOTSUP;
	}
```

Reply:
I don't think that's correct, to me it looks like that would always return after calling `set_public_addr()`.